1X3 ICEBOX | Wilderness Expansion (Now not extremely slow!)

Time-Green opened the pull request:
## About The Pull Request
Turns the surface z-level of icebox into a 1x3 area, effectively adding 2 wilderness new z-levels surrounding the station

Because it's not always clear to other people what I'm talking about, this is what I mean with making the surface level a 3x3 z-level
![image](https://github.com/user-attachments/assets/cb20dcfb-a997-4510-b5ec-25c0622fa944)

(I changed it to 1x3 for performance reasons but it still mostly holds)

Video of me trying to navigate it while the Forested trait randomly ran:
https://discord.com/channels/326822144233439242/326831214667235328/1389352206742061268

The wilderness z-levels are gridlinked, instead of crosslinked, which just means the connections are consistent and not randomized. If you keep going right, you will always end up where you started again, eventually. This also removes the black border around the surface icebox z-level (cause you can just go there now)

**Wilderness levels**
I've added some Z-level templates that can be generated. They're incredibly basic, but all can spawn runes on them as well. 
- Snow planes (5x)
- Ice planes (1x)
- Forest planes (1x)
- Mountain planes (1x)

I've also tweaked surface generation quite a bit. It being completely covered in bones always felt weird, and the intersparsed rocks and chasms never sat right with me. The default overworldgen is now more like the Forested trait, but with more sparse trees. 

All of this is modular btw. You can increase the amount of z-levels, make any space z-level be unrandomized gridlinked or add your own wilderness z-levels (either to your own map or icebox)

<details>
  <summary>Coder details</summary>
  
Forested trait now doesn't spawn in no generation zones, instead using a sparse_generation area to overwrite for nearstation areas

Added new config options for genning the wilderness areas

Made mirages work on non-space tiles

Increased transition edge to 8, since horizontal view is also 8 now and it avoids seeing the border

Will fill out more as I start remembering what I did  
</details>

## ~~Why It's Bad For The Game~~

~~Apparently generating 255 * 255 * 8 extra tiles, all which need to be generates and populated and have light and atmos, has performance costs. I have some bad ideas to speed things up, but I don't see a future where it isn't at least a little slower on roundstart~~

I reduced it to 2 extra z-levels!! Performance cost is significantly reduced. 
![image](https://github.com/user-attachments/assets/4ebed9af-923c-42ee-b45b-de48579279e1)
![image](https://github.com/user-attachments/assets/1c11f49e-f5de-4a42-8f3e-7f885b94a004)


(Normal icebox is around 130 seconds on my PC)


## Why It's Good For The Game
Icebox exploration is kind of depressing. We have this unique setting, but we can't really go anywhere? You can go down and find that one pool, which is about the peak of exploration of icebox.

Now you can literally explore the entire round and get incredibly lost! It's also a great opportunity for mappers! (Especially since the templates I made were made rather quickly as I wasn't sure if this had merit). 

2 extra z-levels isn't a lot, but it'll let us further develop planetary wilderness z-levels further without impacting load times that much. Maybe 3x3 icebox can be real in the future, but for now 1x3 icebox will have to do

## Changelog
:cl:
add: Icebox surface is now a 1x3 z-level grid! Rejoice, explorers!
/:cl:

It can also be easily toggled in the map config, so it can be turned up or down, or different z-level generation can be configured

Time-Green committed:
Makes icebox 3x3

Time-Green committed:
Map improvements

Time-Green committed:
Adds config for modular map selection

Time-Green committed:
Fixes mirages going completely insane

This took so long to figure out

Time-Green committed:
Adds more z-variety

Time-Green committed:
Fixes some active turfs

LemonInTheDark commented:
assertion, even if this doesn't work out we should do this linking on icebox's edges to kill the evil evil border which haunts my nightmares

ArcaneMusic commented:
Immediate thought that requires some additional thought on my part: We'll want to consider custom adjustment of ore generation/vent generation for those adjacent z levels, since it's no surprise at this point that arcmining kinda under populates icebox with ore, it might be worth looking into bumping that up, or giving out a uniform mineral_budget per adjacent surface tile.

uaioy commented:
ice box is about to become even better

jughu commented:
https://youtu.be/xoMLRBCc59Q?si=ANM9_lNTKGJ111ZZ

Maurukas commented:
I did not think I would ever see it, but here we stand, in 2025, at the threshold of PlanetStation.

projectkepler-ru commented:
Space Station 76 looking good lad

Time-Green committed:
Fixes unit test

Krail commented:
Sneak peak of Terry wilderness once this PR passes.
![pkspotsvarrock](https://github.com/user-attachments/assets/1cb52b80-a19f-40fe-aa14-c7548c34c088)
Can't wait for the grind.

tttruancy commented:
Caves of Qud ass pr

LSommers commented:
> Caves of Qud ass pr

hey man don't make fun of his CoQ

jughu commented:
> > > Caves of Qud ass pr
> > 
> > 
> > hey man don't make fun of his CoQ
> 
> I never get past bethesda susa :(

thats only half the fight...

Time-Green committed:
Reduces it to only 2 extra z-levels

spiders24110 commented:
cowards

Time-Green committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into 3x3-icebox-it-is-time

Time-Green committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into 3x3-icebox-it-is-time

Time-Green committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into 3x3-icebox-it-is-time

Time-Green committed:
Redoes my map changes manually, again

LemonInTheDark commented on code/__HELPERS/_lists.dm:
```diff
@@ -528,6 +528,18 @@
+ * A = 5, B = 3, C = 1, D = 0
+ * Tt would then have a 55.55...% to be picked, rinse and repeat
+*/
+/proc/pick_weight_take(list/list_to_pick)
+	. = pick_weight(list_to_pick)
+	list_to_pick[.]--
```
this is a bit... weird, makes actual values matter instead of proportion

Time-Green replied:
I'm not sure I get you? It's so I can define quantities of available options and draw from those without getting duplicates if I don't want to

LemonInTheDark replied:
we normally act like weighted lists only matter in so far as the weights are in relation to the sum, right?
so like, if you de/increase weights by some factor it doesn't change behavior, and this changes this in some cases, s all.

Time-Green replied:
Oh yeah I want it to change the chances, otherwise I'd just have grabbed the normal pickweight

LemonInTheDark commented on code/controllers/subsystem/mapping.dm:
```diff
@@ -132,11 +132,20 @@ SUBSYSTEM_DEF(mapping)
 		empty_space = add_new_zlevel("Empty Area [space_levels_so_far+1]", list(ZTRAIT_LINKAGE = CROSSLINKED))
 		++space_levels_so_far
 
+	if(current_map.wilderness_levels)
+		var/list/FailedZs = list()
+
+		LoadGroup(FailedZs, "Wilderness Area", current_map.wilderness_directory, current_map.maps_to_spawn, default_traits = ZTRAITS_WILDS, height_autosetup = FALSE)
+
+		if(LAZYLEN(FailedZs))
+			CRASH("Ice wilds failed to load!")
```
this is blocked behind the define yes? is that ok?

Time-Green replied:
This is hidden if in LOWMEMORYMODE, yeah. I'm not sure there's any way to detect errors in loading without also loading

LemonInTheDark replied:
I meant the like full load, since these are in theory part of the station. if you're ok with them being optional then it's fine

Time-Green replied:
Oh yeah I'm fine with it. It is pretty chunky. It's why we also refuse to generate lavaland in lowmemorymode so I feel like the precedent is there

LemonInTheDark commented on code/controllers/subsystem/mapping.dm:
```diff
@@ -275,8 +283,7 @@ SUBSYSTEM_DEF(mapping)
 	var/list/ice_ruins = levels_by_trait(ZTRAIT_ICE_RUINS)
 	for (var/ice_z in ice_ruins)
-		var/river_type = HAS_TRAIT(SSstation, STATION_TRAIT_FORESTED) ? /turf/open/lava/plasma/ice_moon : /turf/open/openspace/icemoon
-		spawn_rivers(ice_z, 4, river_type, /area/icemoon/surface/outdoors/unexplored/rivers)
+		spawn_rivers(ice_z, 4, /turf/open/lava/plasma/ice_moon, /area/icemoon/surface/outdoors/unexplored/rivers)
```
what's the deal w this?

LemonInTheDark commented on code/datums/map_config.dm:
```diff
@@ -27,8 +27,15 @@
+
 	/// Boolean that tells us if this is a planetary station. (like IceBoxStation)
 	var/planetary = FALSE
+	/// How many z's to generate around a planetary station
```
around?

LemonInTheDark replied:
ahhh I see

LemonInTheDark commented on code/datums/map_config.dm:
```diff
@@ -27,8 +27,15 @@
 	var/planetary = FALSE
+	/// How many z's to generate around a planetary station
+	var/wilderness_levels = 0
+	/// Directory to the wilderness 0rea we can spawn in
```
```suggestion
	/// Directory to the wilderness area we can spawn in
```

LemonInTheDark commented on code/datums/mapgen/Cavegens/IcemoonCaves.dm:
```diff
@@ -35,12 +35,15 @@
+	weighted_open_turf_types = list(/turf/open/misc/asteroid/snow/icemoon = 1)
 	flora_spawn_chance = 4
 	weighted_mob_spawn_list = null
-	initial_closed_chance = 53
+	initial_closed_chance = 0
```
what's goin on here?

LemonInTheDark replied:
all of them on the top z?

Time-Green replied:
I am a bit biased because I think it also just looks so much better

LemonInTheDark commented on code/game/area/areas/mining.dm:
```diff
@@ -217,8 +217,10 @@
 /// this is the area you use for stuff to not spawn, but if you still want weather.
 /area/icemoon/surface/outdoors/nospawn
 
-// unless you roll forested trait lol (fuck you time green)
```
love you bro

LemonInTheDark commented on code/game/turfs/open/_open.dm:
```diff
@@ -107,6 +113,52 @@
+		var/turf/T = locate(destination_x, destination_y, destination_z)
+		user.forceMove(T)
+
+/turf/open/proc/CanBuildHere()
+	if(destination_z)
+		return FALSE
+	return TRUE
```
why define this like this? shouldn't we at least be using it to prevent building on mirage'd turfs? does this work ok in gameplay?

Time-Green replied:
It DOES prevent building on mirage'd turfs. Don't ask me why, but anytime we check for mirages we check for destination_z. And it works in gameplay. I can't build on it with an RCD and I can't move construction materials there without them being transited.

I don't doubt we'll get some funny issue reports at some point, but there's nothing egregious I can think of

LemonInTheDark commented on code/game/turfs/open/_open.dm:
```diff
@@ -107,6 +113,52 @@
 	. = ..()
 	update_visuals()
 
+//ATTACK GHOST IGNORING PARENT RETURN VALUE
+/turf/open/attack_ghost(mob/dead/observer/user)
+	if(destination_z)
+		var/turf/T = locate(destination_x, destination_y, destination_z)
+		user.forceMove(T)
```
parent call?

Time-Green replied:
Hmm good point. Looks like this would break the ghost gas scan

LemonInTheDark commented on code/game/turfs/open/_open.dm:
```diff
@@ -107,6 +113,52 @@
+		return FALSE
+	return TRUE
+
+/turf/open/space/is_transition_turf()
```
typing is wrong

LemonInTheDark commented on code/game/turfs/open/_open.dm:
```diff
@@ -107,6 +113,52 @@
+	if(!arrived || src != arrived.loc)
+		return
+
+	if(destination_z && destination_x && destination_y && !arrived.pulledby && !arrived.currently_z_moving)
```
early return please

LemonInTheDark commented on code/game/turfs/open/_open.dm:
```diff
@@ -107,6 +113,52 @@
+		var/ty = destination_y
+		var/turf/DT = locate(tx, ty, destination_z)
+		var/itercount = 0
+		while(DT.density || istype(DT.loc,/area/shuttle)) // Extend towards the center of the map, trying to look for a better place to arrive
+			if (itercount++ >= 100)
+				log_game("SPACE Z-TRANSIT ERROR: Could not find a safe place to land [arrived] within 100 iterations.")
+				break
+			if (tx < 128)
+				tx++
+			else
+				tx--
+			if (ty < 128)
+				ty++
+			else
+				ty--
+			DT = locate(tx, ty, destination_z)
```
seems like a poor solution for this particular usecase

LemonInTheDark replied:
yea I know and I don't have a GREAT solution outside of transitions to dense turfs blocking the move, buttt

LemonInTheDark commented on code/modules/mapping/space_management/space_transition.dm:
```diff
@@ -60,39 +60,24 @@
-	var/center = round(grid_diameter / 2)
-	point = grid[CHORDS_TO_1D(grid_diameter, center, center)]
-	grid.Cut()
+		if(linkage)
+			if(level.linkage != linkage)
+				// Either you be gridlinked or crosslinked, both is uhhh... a headache
+				stack_trace("Mixed linkages detected in z-level neighbour transitions!")
+		else
+			linkage = level.linkage
```
```suggestion
		if(!isnull(linkage) && level.linkage != linkage)
			// Either you be gridlinked or crosslinked, both is uhhh... a headache
			stack_trace("Mixed linkages detected in z-level neighbour transitions!")
			continue
		linkage = level.linkage
```
confused me reading it, didn't realize it was a guard case.
also, should we say which level disagrees? to help mappers

LemonInTheDark commented on code/modules/mapping/space_management/space_transition.dm:
```diff
@@ -162,4 +147,61 @@
 				S.AddElement(/datum/element/mirage_border, place, mirage_dir, range_cached)
 
+/// Construct linkages randomly to get confusing space transitions
+/// We do this by constructing an absurdly large grid, and placing the levels randomly inside
+/// Please update this comment when you figure out why this is
```
can you elaborate on what part of this bugs you, I at once point understood this code so I might be able to help

Time-Green replied:
I actually kind of understand it now too. I should swap this comment out with something more useful

LemonInTheDark commented on code/modules/mapping/space_management/space_transition.dm:
```diff
@@ -162,4 +147,61 @@
+	var/datum/space_transition_point/point
+	for(var/x in 1 to grid_diameter)
+		for(var/y in 1 to grid_diameter)
+			point = new/datum/space_transition_point(x, y, grid)
```
```suggestion
			point = new /datum/space_transition_point(x, y, grid)
```

LemonInTheDark commented on code/modules/mapping/space_management/space_transition.dm:
```diff
@@ -162,4 +147,61 @@
+	var/datum/space_transition_point/point
+	for(var/x in 1 to grid_diameter)
+		for(var/y in 1 to grid_diameter)
+			point = new/datum/space_transition_point(x, y, grid)
```
```suggestion
			point = new /datum/space_transition_point(x, y, grid)
```
ikik

LemonInTheDark commented on code/modules/mapping/space_management/space_transition.dm:
```diff
@@ -162,4 +147,61 @@
+		level.xi = point.x
+		level.yi = point.y
+		point.spl = level
+		used_points += point //this used_points list is kinda lame, you can remove it if you can find out what the slice function in byond is
```
Copy -> Cut unfortunately, this might just be better idk what you would want to do here

Time-Green replied:
Such is life

LemonInTheDark commented on code/__DEFINES/maps.dm:
```diff
@@ -44,7 +44,7 @@ Always compile, always use that verb, and always make sure that it works for wha
 
 /// Distance from edge to move to another z-level
-#define TRANSITIONEDGE 7
+#define TRANSITIONEDGE 8
```
speak on this

LemonInTheDark replied:
also should it be like 9 or whatever cause fullscreen

Time-Green replied:
Yeah cause uhh we increased the screen width a while ago and never updated this. It doesn't really stand out when you're in space because it's black on black, but it does in icebox.

Also width transition edge is +1, so this is the correct value for the define

<img width="740" height="280" alt="image" src="https://github.com/user-attachments/assets/070eca2c-8931-4f04-893f-efc45fcff81b" />

Time-Green committed:
Update code/datums/map_config.dm

Co-authored-by: LemonInTheDark <58055496+LemonInTheDark@users.noreply.github.com>

Time-Green committed:
Update code/modules/mapping/space_management/space_transition.dm

Co-authored-by: LemonInTheDark <58055496+LemonInTheDark@users.noreply.github.com>

Time-Green committed:
Apply suggestions from code review

Co-authored-by: LemonInTheDark <58055496+LemonInTheDark@users.noreply.github.com>

Time-Green committed:
Review changes

Time-Green committed:
Merge branch '3x3-icebox-it-is-time' of https://github.com/Time-Green/tgstation into 3x3-icebox-it-is-time

Time-Green committed:
Generalizes CanBuildHere

MrMelbert commented:
Also the complete lack of walls on any upper z-level... really doesn't look good - There's not much value in having a full z-level which is 100% flat barring the occasional rock or tree and the even less occasional ruin

<img width="1116" height="1012" alt="image" src="https://github.com/user-attachments/assets/b994e159-e180-446e-a972-6a74b2e83a45" />

A lot of these ruins were designed to be slotted into surrounding rock and they sure stick out 

<img width="551" height="870" alt="image" src="https://github.com/user-attachments/assets/05cda9e7-7641-46ee-bfe2-8aca85d5f8f9" />

MrMelbert commented:
I'm 100% unopposed to the wilderness levels but they either
- Need some walls
Or
- We can add walls via ruin templates, the same way we add asteroids to space via ruin templates: IE, a ruin which is literally nothing but a rock formation, no loot or anything.
- Also would need to strip walls from existing ruins

Time-Green commented:
> I'm 100% unopposed to the wilderness levels but they either
> 
> * Need some walls
> * And maybe up the foliage density slightly? At least grass if not trees
> * And also perhaps a few more rivers
> 
> Or
> 
> * We can add walls via ruin templates, the same way we add asteroids to space via ruin templates: IE, a ruin which is literally nothing but a rock formation, no loot or anything.
> * Also would need to strip walls from existing ruins

I've changed the terrain generator a bit. I've increased foliage, and changed the default template to rocky_planes.dmm which has a more maze-like layout

<img width="1355" height="944" alt="image" src="https://github.com/user-attachments/assets/3697179e-8e94-4f5e-964e-6d7785e15287" />

There is still a small chance for the wall-less snow planes to spawn (1/16 per z-level spawn). I could spawn some rocks on the station z-level, but it's limited by the new transition border and wont look very good. 

The "strange effects" you saw is a glitch in how the transit borders work, it just rarely stands out in normal space and requires you to stand on a diagonal. I've tried patching it but it's not an easy fix

In concern to the ruins sticking out: I'd prefer this to actually get merged before I change every icebox ruin. I guess it's also less pressing now that it's mostly wall again. I was gonna work with Kryson to get some more fluff in (🗿), but yeah uhhhhhhh after this is merged

Time-Green committed:
More rockyness

Time-Green committed:
Increases river density by 50%

Time-Green committed:
Tones down mob spawns in rocky planes

Time-Green committed:
Wawa

MrMelbert commented:
> I've changed the terrain generator a bit. I've increased foliage, and changed the default template to rocky_planes.dmm which has a more maze-like layout. I have also increased river density by 50%

Looks good to me

> In concern to the ruins sticking out: I'd prefer this to actually get merged before I change every icebox ruin. I guess it's also less pressing now that it's mostly wall again. I was gonna work with Kryson to get some more fluff in (🗿), but yeah uhhhhhhh after this is merged

Oh don't get me confused I only said that they'd stick badly out if there were 0 rock walls, and you added rock walls, so they're not going to stick out badly anymore
